### PR TITLE
Add Windows support

### PR DESCRIPTION
**Pull request by marlenecota**
# Summary

Explain the **motivation** for making this change: here are some points to help you:

* What issues does the pull request solve? Please tag them so that they will get automatically closed once the PR is merged
#297
#1426
#1610
* What is the feature? (if applicable)
Adding support for react-native-windows. 
Done: Svg, Rect, Circle, Ellipse, Line, Polygon, Polyline, Path, G, Use, Symbol, Defs, Image, LinearGradient, RadialGradient, Pattern
Partial: Text, TSpan
Missing: TextPath, ClipPath, Mask, Marker, ForeignObject. 
* How did you implement the solution?
Using [Win2D](https://github.com/microsoft/Win2D)
* What areas of the library does it impact?

## Test Plan

Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.

https://user-images.githubusercontent.com/1422161/129789622-a39837d3-1d29-4d16-a552-ea280de60cb1.mp4

https://user-images.githubusercontent.com/1422161/129789657-3fa54117-cbc6-4635-b3d9-a11b98d4f0e1.mp4

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ❌     |
| Windows |    ✅     |

## Checklist

- [x] I have tested this on a device and a simulator
- [ ] I added documentation in `README.md`
- [ ] I updated the typed files (typescript)
- [ ] I added a test for the API in the `__tests__` folder

### CODE REVIEW

File: windows/RNSVG/PathView.cpp
Comment (PPatBoyd) [inline windows/RNSVG/PathView.cpp:48]: Is there a longer plan of having this parsing supported by Graphics.Canvas.Svg?  Having to do this parsing here feels rough

**Review by PPatBoyd — COMMENTED** [review]
I noticed clipPath / clipRule aren't implemented, are those the only props not yet implemented on the elements marked as done?

Does d2d1_fill_mode alternate/winding enable the fill and clip rules `evenodd` and `nonzero` correctly?   [https://docs.microsoft.com/en-us/windows/win32/api/d2d1/ne-d2d1-d2d1_fill_mode?redirectedfrom=MSDN alternating/winding ](D2D1_FILL_MODE)
 
Are there any props or elements (e.g. Mask) that we don't intend to implement?  As a point of comparison d2d1 SVG doesn't implement a Mask control, not sure what issues that presents

File: windows/RNSVG/GroupView.cpp
Comment (marlenecota) [reply windows/RNSVG/GroupView.cpp:99]: Invalidate is false when the prop value is not being set on the element currently being updated. For example, if you set the fill on a G, then that group passes that value on to all its children. In that case, we don't want to invalidate/redraw the canvas every time a child is updated, so we pass false. Once it is done propagating, then the G element calls Invalidate once to redraw.

File: windows/RNSVG/PathView.cpp
Comment (marlenecota) [reply windows/RNSVG/PathView.cpp:48]: While investigating how to implement this I came upon [this](https://github.com/microsoft/Win2D/issues/186), but I haven't reached out to the Win2D to see if they have changed their minds. 

This parse code / solution is taken from the [Android implementation](https://github.com/react-native-svg/react-native-svg/blob/develop/android/src/main/java/com/horcrux/svg/PathParser.java), so we at least have parity there.

**Comment by marlenecota** [comment]
> I noticed clipPath / clipRule aren't implemented, are those the only props not yet implemented on the elements marked as done?

I am on vacation through Monday, but I will update the PR with a more complete feature parity table when I get back.

> 
> Does d2d1_fill_mode alternate/winding enable the fill and clip rules `evenodd` and `nonzero` correctly? [https://docs.microsoft.com/en-us/windows/win32/api/d2d1/ne-d2d1-d2d1_fill_mode?redirectedfrom=MSDN alternating/winding ](D2D1_FILL_MODE)

The way fill rule is currently implemented, it indirectly uses D2D1_FILL_MODE via Canvas Geometry.CreateGroup and it looks to be working as expected. I'll look into using that for clip. 

> Are there any props or elements (e.g. Mask) that we don't intend to implement? As a point of comparison d2d1 SVG doesn't implement a Mask control, not sure what issues that presents

Because we're creating Win2D geometry objects, we're able to implement elements and props not supported by d2d1 SVG (the style prop being one example). It might not come all at once, but the goal is to have feature parity with iOS and Android, so nothing is off the table yet.

File: windows/RNSVG/GroupView.cpp
Comment (asklar) [inline windows/RNSVG/GroupView.cpp:48]: could default to Segoe UI Variable for Windows 11+

File: windows/RNSVG/GroupView.h
Comment (asklar) [inline windows/RNSVG/GroupView.h:52]: use `std::unordered_map` for small maps

File: windows/RNSVG/ImageView.cpp
Comment (asklar) [inline windows/RNSVG/ImageView.cpp:236]: should at least yellowbox?

File: windows/RNSVG/ImageView.cpp
Comment (asklar) [inline windows/RNSVG/ImageView.cpp:262]: yellowbox?

File: windows/RNSVG/ImageView.h
Comment (asklar) [inline windows/RNSVG/ImageView.h:9]: can this be a Uri object?

File: windows/RNSVG/ImageViewManager.cpp
Comment (asklar) [inline windows/RNSVG/ImageViewManager.cpp:25]: these should be numbers, no?

File: windows/RNSVG/ImageViewManager.cpp
Comment (asklar) [inline windows/RNSVG/ImageViewManager.cpp:30]: is this a number or a bool?

File: windows/RNSVG/LineViewManager.cpp
Comment (asklar) [inline windows/RNSVG/LineViewManager.cpp:25]: Number

Comment (asklar) [reply windows/RNSVG/LineViewManager.cpp:25]: looks like there are other cases like this so please do a pass over them :)

File: windows/RNSVG/SvgView.cpp
Comment (asklar) [inline windows/RNSVG/SvgView.cpp:22]: exceptions that come from CreateGeometry (e.g. a parsing error) would end up bubbling to this lambda, and probably killing the app. Can you add some error handling here to at least redbox/yellowbox?

File: windows/RNSVG/PathView.cpp
Comment (asklar) [inline windows/RNSVG/PathView.cpp:32]: @mcota I think we don't need any of the parsing logic and instead can use `XamlBindingHelper.ConvertValue`(https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.markup.xamlbindinghelper.convertvalue?view=winrt-20348)

File: windows/RNSVG/PathView.cpp
Comment (asklar) [reply windows/RNSVG/PathView.cpp:48]: I am hoping you can delete all of the parsing code and just use XamlBindingHelper instead

File: windows/RNSVG/ImageViewManager.cpp
Comment (marlenecota) [reply windows/RNSVG/ImageViewManager.cpp:25]: No, they're SVGLength values that come through as a string

File: windows/RNSVG/ImageViewManager.cpp
Comment (marlenecota) [reply windows/RNSVG/ImageViewManager.cpp:30]: It's a [number](https://github.com/react-native-svg/react-native-svg/blob/develop/src/elements/Image.tsx#L52). meet is 0, slice is 1, none is 2.

File: windows/RNSVG/LineViewManager.cpp
Comment (marlenecota) [reply windows/RNSVG/LineViewManager.cpp:25]: Same as the others. These are SVGLength values not just a number.

File: windows/RNSVG/Utils.h
Comment (Saadnajmi) [reply windows/RNSVG/Utils.h:313]: Curious, does this PR mean PlatformColor as colors would be supported on windows for RN-SVG?

**Comment by kingofspartacus** [comment]
Hi @marlenecota thanks you for sharing, currently I am having this problem when build release mode
image

**Comment by chrisglein** [comment]
@marlenecota Ready to move this out of draft? What more is this waiting for?

File: windows/RNSVG/Utils.h
Comment (marlenecota) [reply windows/RNSVG/Utils.h:313]: This implementation is set up to support PlatformColor, but it will not work right now for the same reason it does not currently work in other platforms - the javascript side of the module is doing its own [color value parsing](https://github.com/react-native-svg/react-native-svg/blob/develop/src/lib/extract/extractColor.ts) instead of using react-native's [processColor](https://github.com/facebook/react-native/blob/main/Libraries/StyleSheet/processColor.js).

**Comment by true-hamid** [comment]
We have tested this and it's working fine in our app, is there an expected date for this pr to be completed?

**Comment by chrisglein** [comment]
> We have tested this and it's working fine in our app, is there an expected date for this pr to be completed?

Looks like it's waiting on an official sign off. @marlenecota, still looking for reviews?

**Comment by rafi16jan** [comment]
My project currently uses React Native on all platforms even on the web, except for desktop (still uses electron), this thing is a life saver because we already planned to port the electron build to React Native Windows + MacOS and 80% of our UI components are made of SVGs :')

**Comment by kingofspartacus** [comment]
> > Hi @marlenecota thanks you for sharing, currently I am having this problem when build release mode image
> 
> @kingofspartacus are you still hitting this? I've tried this on several machines and I'm not seeing it.

Thank you for your reply, Currently, it works fine after I install the package Win2D.uwp 
image_2021_11_19T15_00_34_663Z

**Comment by x-ji** [comment]
Just curious: while Googling around I saw that MS announced that they started to support SVG natively with React Native for Windows in this post dating to Feb 2020. https://microsoft.github.io/react-native-windows/blog/2020/02/010/m4updates#svg-image-support I guess this library probably provides more flexibility and configuration options?

**Comment by marlenecota** [comment]
> Just curious: while Googling around I saw that MS announced that they started to support SVG natively with React Native for Windows in this post from Feb 2020. https://microsoft.github.io/react-native-windows/blog/2020/02/010/m4updates#svg-image-support I guess this library probably provides more flexibility and configuration options? Or I guess maybe a number of existing libraries already depend on this library for them to work.

All of the above :)

Also, the SVG support added to the Image component is also limited by Direct2D:
https://docs.microsoft.com/en-us/windows/win32/direct2d/svg-support

For example, style prop/attribute does not work when loading an svg file using the Image component, but it does work when using react-native-svg.

**Comment by SBS-FPATE-RINALDI** [comment]
I tried referencing this PR directly in my package.json and it didn't work on Windows. I noticed the new Windows folder was not referenced in the list of included files in the package.json. Doesn't it need to be included there or am I missing something?

**Comment by guillemonzo** [comment]
Any updates on this?

**Comment by jdk** [comment]
@PPatBoyd @marlenecota can this be merged? 

@marlenecota you mentioned a parity document, was wondering if I missed this?

**Comment by markokhman** [comment]
Did anybody try to merge this?

**Comment by markokhman** [comment]
Guys, please let me know what is this issue up to? I can contribute by testing and debugging.
Seems like this is about to be ready for merging.

**Comment by jdk** [comment]
@msand We've tested this too. Is there anything we can further do to help expedite a merge?

**Comment by shanelile** [comment]
I'd also like to know what the status here is - is there anything folks can do to help out with this?

**Comment by marlenecota** [comment]
> I'd also like to know what the status here is - is there anything folks can do to help out with this?
Text, TSpan, TextPath, ClipPath, Mask, Marker, and ForeignObject have been put on the back burner, but this could be merged as is for a partial implementation. Need someone with write access to be ok with that and approve it.

**Comment by ryanlntn** [comment]
@marlenecota I'm seeing this when trying to build for ARM64

Screen Shot 2022-06-15 at 8 55 38 AM

I tried running NuGet package restore but still get the same error. Any idea what I might be missing?

**Comment by alexzobi** [comment]
@Saadnajmi @asklar @PPatBoyd sorry to be annoying, but would it be possible to take a peak at this? Even this partial implementation would be a huge boon for me and it looks like many others. @marlenecota thanks for all of the hard work!

**Comment by Saadnajmi** [comment]
> @Saadnajmi @asklar @PPatBoyd sorry to be annoying, but would it be possible to take a peak at this? Even this partial implementation would be a huge boon for me and it looks like many others. @marlenecota thanks for all of the hard work!

Software Mansion maintains React Native SVG now, so I think the person who could comment on whether this pointed up with their future plans for the repo is @WoLewicki . I know they were working on bringing Fabric support.

**Comment by gillpeacegood** [comment]
@marlenecota it seems it could be merged back?

**Comment by jdk** [comment]
@marlenecota 👍 

Tested and it’s ready. Please. It’s ready.

**Comment by taduyde** [comment]
ready to merge?

**Comment by marlenecota** [comment]
@jdk - which version of RN/RNW did you test?

This was created and tested in 0.64. I am currently spread too thin with other work to validate newer versions. If anyone has been able to validate this change against newer versions or can volunteer time to do so, please report results back on this thread. 

If we want to go ahead with the merge, any upgrading issues could be addressed in future PRs.

**Comment by jdk** [comment]
> @jdk - which version of RN/RNW did you test?
> 
> This was created and tested in 0.64. I am currently spread too thin with other work to validate newer versions. If anyone has been able to validate this change against newer versions or can volunteer time to do so, please report results back on this thread.
> 
> If we want to go ahead with the merge, any upgrading issues could be addressed in future PRs.
 
"react-native": "0.67.4",
"react-native-windows": "0.67.7"

**Comment by jeanmaried** [comment]
Working on:
    "react-native": "0.67.2",
    "react-native-windows": "0.67.7"

**Comment by WoLewicki** [comment]
@marlenecota if you are OK with merging it, we can do it and then release a new version of library so people can test it thoroughly.

**Comment by lcherukuri** [comment]
When can we expect this to merge and release?

**Comment by marlenecota** [comment]
> @marlenecota if you are OK with merging it, we can do it and then release a new version of library so people can test it thoroughly.

@WoLewicki sound good!

File: windows/RNSVG/Utils.h
Comment (WoLewicki) [inline windows/RNSVG/Utils.h:313]: So since https://github.com/react-native-svg/react-native-svg/pull/1726 was merged, it should work correctly then @marlenecota ?

**Review by WoLewicki — APPROVED** [review]
Great job! Let's merge it 🚀

**Pull request merged**
